torch/utils/data/sampler.py

```diff
@@ -83,7 +83,7 @@ class RandomSampler(Sampler[int]):
     replacement: bool
 
     def __init__(self, data_source: Sized, replacement: bool = False,
-                 num_samples: Optional[int] = None, generator=None) -> None:
+                 num_samples: int = None, generator=None) -> None:
```

Comment:
if its type is `int` and if kept optional argument, should its default value be changed to `0`? (the guard condition currently asserts that num_samples must be greater than zero. btw supporting fully empty datasets may also be useful in some cases as is supporting empty batches)

Reply:
Thank you for your comment - I think that its value is meant to be defaulted to `None`. Something I just realized is that the guard condition does its assertion on the `num_samples` property, not the `_num_samples` attribute.

The `_num_samples `attribute is set to the argument of `__init__`:
https://github.com/pytorch/pytorch/blob/1c5a8125798392f8d7c57e88735f43a14ae0beca/torch/utils/data/sampler.py#L89

The `num_samples `property defaults to the length of the data source upon retrieval:
https://github.com/pytorch/pytorch/blob/1c5a8125798392f8d7c57e88735f43a14ae0beca/torch/utils/data/sampler.py#L100-L105

Because of this, I believe the num_samples argument should be kept as `Optional[int]`, with a default value of `None`. Let me know if you have any other thoughts, otherwise I will be closing this issue.